### PR TITLE
SageMath wrappers

### LINKED ISSUE #64: Introduce SageMath parents & elements
see https://github.com/videlec/sage-flatsurf/issues/12.

### PR DESCRIPTION
**Pull request by saraedum**
fixes #64

Some design notes:
* Pickling goes through cereal, i.e., we turn our objects into JSON and then dump that JSON string to pickle things. We could be a bit more efficient by using  the binary cereal format but I prefer readable pickles for the time being. More importantly, this does not properly deduplicate the pickles: If we pickle `L=[x, x.parent()]`, then `x` also pickles a separate copy of `x.parent()` which bloats the pickle somewhat. It's conceivable that one could hook into the pickling mechanisms to drop this duplicate but it's a lot of work. Note, however, that after unpickling the duplicate is dropped again, i.e., `L[0].parent() is L[1]`.
* Calling `.module()` on an element returns a `shared_ptr<Module<…>>`. We don't get `x.module() is x.module()` since two shared pointers are not identical here. The object these pointers point to are identical of course.
* `ExactReals` are not in `Fields()` but in `IntegralDomains()`. Since we essentially have no inverses, `Fields()` causes too much trouble.

### CODE REVIEW

**Comment by saraedum** [comment]
@videlec, feel free to have a look. No need for an actual review here I think. I point out some interesting bits in the comments…

File: pyexactreal/src/pyexactreal/__init__.py [hunk 1]
```diff
@@ -15,6 +25,74 @@
...
-#*********************************************************************
+# ********************************************************************
+
+from .cppyy_exactreal import exactreal, ZZModule, QQModule, NumberFieldModule
+
+import pyeantic
+NumberField = pyeantic.eantic.renf
+RealNumber = exactreal.RealNumber
+
+def ExactReals(base=None, *args, **kwargs):
```
Comment (saraedum) [inline pyexactreal/src/pyexactreal/__init__.py]: @videlec that's the interface at the moment.

File: pyexactreal/src/pyexactreal/cppyy_exactreal.py [hunk 2]
```diff
@@ -29,76 +75,152 @@
...
 cppyy.py.add_pythonization(enable_arithmetic, "exactreal")
+cppyy.py.add_pythonization(enable_arithmetic, "exactreal::yap")
+
+def enable_yap(proxy, name):
+    r"""
+    Make yap templating machinery accessible from cppyy.
+
+    TESTS::
+
+        >>> from pyexactreal import exactreal
```
Comment (saraedum) [inline pyexactreal/src/pyexactreal/cppyy_exactreal.py]: @videlec this also fixes Arb from Python I hope. Before I had some spurious problems with segfaults. Also the printing should now make it clear what's going on.

**Comment by videlec** [comment]
I can not compile
```
In file included from /usr/include/c++/9.1.0/bits/move.h:55,
                 from /usr/include/c++/9.1.0/bits/stl_pair.h:59,
                 from /usr/include/c++/9.1.0/bits/stl_algobase.h:64,
                 from /usr/include/c++/9.1.0/vector:60,
                 from /opt/sage/local/include/e-antic/renfxx.h:17,
                 from element.cc:21:
/usr/include/c++/9.1.0/type_traits: In substitution of 'template<bool _Cond, class _Tp> using enable_if_t = typename std::enable_if::type [with bool _Cond = exactreal::IntegerRing::isField; _Tp = void]':
exact-real/element.hpp:74:12:   required from 'class exactreal::Element<exactreal::IntegerRing>'
element.cc:465:27:   required from here
/usr/include/c++/9.1.0/type_traits:2426:11: error: no type named 'type' in 'struct std::enable_if<false, void>'
 2426 |     using enable_if_t = typename enable_if<_Cond, _Tp>::type;
      |           ^~~~~~~~~~~
compilation terminated due to -fmax-errors=1.
make[3]: *** [Makefile:638: element.lo] Error 1
```

File: pyexactreal/src/pyexactreal/__init__.py [hunk 3]
```diff
@@ -15,6 +25,74 @@
...
+
+    Internally, elements live in different modules, generated by the generators
+    their built from. However, you won't see any of that underlying machinery
+    as these modules get constructed automatically as required::
+
+        sage: g*g + g + a
+        ℝ(0.178808…)*ℝ(0.178808…) + ℝ(0.178808…) + (a ~ 1.414214)
+
+    Basic arithmetic is supported, though some operations, such as divisions,
+    only work in special cases::
```
[the remark this answers is not in the document]
Comment (saraedum) [reply pyexactreal/src/pyexactreal/__init__.py]: Sure. It's a subclass of ArithmeticError.

File: pyexactreal/src/pyexactreal/__init__.py [hunk 4]
```diff
@@ -15,6 +25,74 @@
...
+        ℝ(0.178808…) + 1
+        sage: a + g
+        (a ~ 1.414214) + ℝ(0.178808…)
+
+    Internally, elements live in different modules, generated by the generators
+    their built from. However, you won't see any of that underlying machinery
+    as these modules get constructed automatically as required::
+
+        sage: g*g + g + a
+        ℝ(0.178808…)*ℝ(0.178808…) + ℝ(0.178808…) + (a ~ 1.414214)
```
[the remark this answers is not in the document]
Comment (saraedum) [reply pyexactreal/src/pyexactreal/__init__.py]: Ok. Fixed.

**Comment by saraedum** [comment]
@videlec: Finally an e-antic conflict:
```
Package e-antic conflicts for:
e-antic=1
pyeantic -> e-antic[version='1.*|>=1.0a,<2.0a0']
sage -> gap[version='>=4.8.6,<5'] -> normaliz[version='>=3.7.4,<4.0a0'] -> e-antic[version='>=0.1.3,<1.0a0']
```

It's easy to work around as we do not need the full SageMath distribution.

**Comment by videlec** [comment]
Sadly, we will stay with this conflict until @w-bruns make the switch to e-antic 1.0 in Normaliz.

**Comment by videlec** [comment]
Don't we want some uniqueness of the modules
```
sage: from pyexactreal import ExactReals
sage: R = ExactReals(); R
Real Numbers as (Rational Field)-Module
sage: R(1).module() is R(1).module()
False
```
similarly with
```
sage: g = R.random_element()
sage: g.module() is g.module()
False
```

**Comment by videlec** [comment]
Sage test suite would better pass
```
sage: TestSuite(ExactReals()).run()
...
The following tests failed: _test_elements, _test_quo_rem
```
and
```
sage: x = polygen(QQ)
sage: root = AA.polynomial_root(x^3 - 2 * x^2 - 3* x - 1, RIF(3.07, 3.08))
sage: K.<a> = NumberField(x^3 - 2 * x^2 - 3* x - 1, embedding=root)
sage: TestSuite(ExactReals(K)).run()
...
The following tests failed: _test_elements, _test_one, _test_prod, _test_quo_rem
```

**Comment by videlec** [comment]
(wishlist) Given real numbers `R1` as a `K1`-module and `R2` as `K2`-module with `K_1 c K_2` don't we want coercion `R1 -> R2`? In other words
```
sage: R1 = ExactReals()
sage: x = polygen(QQ)
sage: K.<sqrt2> = NumberField(x^2 - 2, embedding=AA(2).sqrt())
sage: R2 = ExactReals(K)
sage: R2.has_coerce_map_from(R1)   # I would like this to be true
False
```

**Comment by videlec** [comment]
Serious bug
```
sage: K.<sqrt2> = NumberField(x^2 - 2, embedding=AA(2).sqrt())
sage: E = ExactReals(K)
sage: E(0) * E(1)
...
Exception: Could not instantiate mul<exactreal::Element<exactreal::NumberField>,exactreal::Element<exactreal::NumberField>>:
  none of the 5 overloaded methods succeeded. Full details:
  exactreal::Element<exactreal::NumberField> exactreal::cppyy::mul(const exactreal::Element<exactreal::NumberField>& lhs, const exactreal::Element<exactreal::NumberField>& rhs) =>
    Exception: not implemented - multiplication in modules over different rings (C++ exception)
  exactreal::Element<exactreal::NumberField> exactreal::cppyy::mul(const exactreal::Element<exactreal::NumberField>& lhs, const exactreal::Element<exactreal::NumberField>& rhs) =>
    Exception: not implemented - multiplication in modules over different rings (C++ exception)
  exactreal::Element<exactreal::NumberField> exactreal::cppyy::mul(const exactreal::Element<exactreal::NumberField>& lhs, const exactreal::Element<exactreal::NumberField>& rhs) =>
    Exception: not implemented - multiplication in modules over different rings (C++ exception)
  exactreal::Element<exactreal::NumberField> exactreal::cppyy::mul(const exactreal::Element<exactreal::NumberField>& lhs, const exactreal::Element<exactreal::NumberField>& rhs) =>
    Exception: not implemented - multiplication in modules over different rings (C++ exception)
  exactreal::Element<exactreal::NumberField> exactreal::cppyy::mul(const exactreal::Element<exactreal::NumberField>& lhs, const exactreal::Element<exactreal::NumberField>& rhs) =>
    Exception: not implemented - multiplication in modules over different rings (C++ exception)
  none of the 6 overloaded methods succeeded. Full details:
  exactreal::Element<exactreal::NumberField> exactreal::cppyy::mul(const exactreal::Element<exactreal::NumberField>& lhs, const exactreal::Element<exactreal::NumberField>& rhs) =>
    Exception: not implemented - multiplication in modules over different rings (C++ exception)
  exactreal::Element<exactreal::NumberField> exactreal::cppyy::mul(const exactreal::Element<exactreal::NumberField>& lhs, const exactreal::Element<exactreal::NumberField>& rhs) =>
    Exception: not implemented - multiplication in modules over different rings (C++ exception)
  exactreal::Element<exactreal::NumberField> exactreal::cppyy::mul(const exactreal::Element<exactreal::NumberField>& lhs, const exactreal::Element<exactreal::NumberField>& rhs) =>
    Exception: not implemented - multiplication in modules over different rings (C++ exception)
  exactreal::Element<exactreal::NumberField> exactreal::cppyy::mul(const exactreal::Element<exactreal::NumberField>& lhs, const exactreal::Element<exactreal::NumberField>& rhs) =>
    Exception: not implemented - multiplication in modules over different rings (C++ exception)
  exactreal::Element<exactreal::NumberField> exactreal::cppyy::mul(const exactreal::Element<exactreal::NumberField>& lhs, const exactreal::Element<exactreal::NumberField>& rhs) =>
    Exception: not implemented - multiplication in modules over different rings (C++ exception)
  exactreal::Element<exactreal::NumberField> exactreal::cppyy::mul(const exactreal::Element<exactreal::NumberField>& lhs, const exactreal::Element<exactreal::NumberField>& rhs) =>
    Exception: not implemented - multiplication in modules over different rings (C++ exception)
```

**Comment by videlec** [comment]
Tiny patch suggestion
```
diff --git a/pyexactreal/src/pyexactreal/exact_reals.py b/pyexactreal/src/pyexactreal/exact_reals.py
index ee351f5..e2aef71 100644
--- a/pyexactreal/src/pyexactreal/exact_reals.py
+++ b/pyexactreal/src/pyexactreal/exact_reals.py
@@ -293,7 +293,7 @@ class ExactReals(UniqueRepresentation, Field):
 
         """
         base = base or QQ
-        category = category or Fields(base)
+        category = category or Fields(base).Infinite()
         return super(ExactReals, cls).__classcall__(cls, base, category)
 
     def __init__(self, base=None, category=None):
```

**Comment by videlec** [comment]
(wishlist) It would be very nice to have conversion of exact reals to real interval field and real ball field. That is obtained by implementing `_real_mpfi_` and `_arb_` (see for example `sage/rings/number_field/number_field_element_quadratic.pyx`).

**Comment by w-bruns** [comment]
I am in a dilemma: I think e-antic 1.0 requires C++17, but many of my colleagues agree that it is not a good idea to change to C++17 already (?) now. It could be a compromise to set e-antic 1.0 back to C++ 14. This is only a matter of code aesthetics, I think, not of functionality or performance.

CC @fingolfin

File: pyexactreal/src/pyexactreal/exact_reals.py [hunk 5]
```diff
@@ -0,0 +1,435 @@
...
+
+            sage: from pyexactreal import ExactReals
+            sage: R = ExactReals()
+            sage: R.random_element()
+            ℝ(...)
+
+        """
+        return repr(self._backend)
+
+    def _cmp_(self, rhs):
```
[the remark this answers is not in the document]
Comment (saraedum) [reply pyexactreal/src/pyexactreal/exact_reals.py]: https://github.com/flatsurf/exact-real/issues/68

**Comment by saraedum** [comment]
> sage: from pyexactreal import ExactReals
> sage: R = ExactReals(); R
> Real Numbers as (Rational Field)-Module
> sage: R(1).module() is R(1).module()

Ok. I added a test.

**Comment by saraedum** [comment]
> Sage test suite would better pass
> 
> ```
> sage: TestSuite(ExactReals()).run()
> ...
> The following tests failed: _test_elements, _test_quo_rem
> ```
> 
> and
> 
> ```
> sage: x = polygen(QQ)
> sage: root = AA.polynomial_root(x^3 - 2 * x^2 - 3* x - 1, RIF(3.07, 3.08))
> sage: K.<a> = NumberField(x^3 - 2 * x^2 - 3* x - 1, embedding=root)
> sage: TestSuite(ExactReals(K)).run()
> ...
> The following tests failed: _test_elements, _test_one, _test_prod, _test_quo_rem
> ```

Also added a test for this.

**Comment by saraedum** [comment]
> Serious bug
> 
> ```
> sage: K.<sqrt2> = NumberField(x^2 - 2, embedding=AA(2).sqrt())
> sage: E = ExactReals(K)
> sage: E(0) * E(1)
> ...
> Exception: Could not instantiate mul<exactreal::Element<exactreal::NumberField>,exactreal::Element<exactreal::NumberField>>:
>   none of the 5 overloaded methods succeeded. Full details:
>   exactreal::Element<exactreal::NumberField> exactreal::cppyy::mul(const exactreal::Element<exactreal::NumberField>& lhs, const exactreal::Element<exactreal::NumberField>& rhs) =>
>     Exception: not implemented - multiplication in modules over different rings (C++ exception)
>   exactreal::Element<exactreal::NumberField> exactreal::cppyy::mul(const exactreal::Element<exactreal::NumberField>& lhs, const exactreal::Element<exactreal::NumberField>& rhs) =>
>     Exception: not implemented - multiplication in modules over different rings (C++ exception)
>   exactreal::Element<exactreal::NumberField> exactreal::cppyy::mul(const exactreal::Element<exactreal::NumberField>& lhs, const exactreal::Element<exactreal::NumberField>& rhs) =>
>     Exception: not implemented - multiplication in modules over different rings (C++ exception)
>   exactreal::Element<exactreal::NumberField> exactreal::cppyy::mul(const exactreal::Element<exactreal::NumberField>& lhs, const exactreal::Element<exactreal::NumberField>& rhs) =>
>     Exception: not implemented - multiplication in modules over different rings (C++ exception)
>   exactreal::Element<exactreal::NumberField> exactreal::cppyy::mul(const exactreal::Element<exactreal::NumberField>& lhs, const exactreal::Element<exactreal::NumberField>& rhs) =>
>     Exception: not implemented - multiplication in modules over different rings (C++ exception)
>   none of the 6 overloaded methods succeeded. Full details:
>   exactreal::Element<exactreal::NumberField> exactreal::cppyy::mul(const exactreal::Element<exactreal::NumberField>& lhs, const exactreal::Element<exactreal::NumberField>& rhs) =>
>     Exception: not implemented - multiplication in modules over different rings (C++ exception)
>   exactreal::Element<exactreal::NumberField> exactreal::cppyy::mul(const exactreal::Element<exactreal::NumberField>& lhs, const exactreal::Element<exactreal::NumberField>& rhs) =>
>     Exception: not implemented - multiplication in modules over different rings (C++ exception)
>   exactreal::Element<exactreal::NumberField> exactreal::cppyy::mul(const exactreal::Element<exactreal::NumberField>& lhs, const exactreal::Element<exactreal::NumberField>& rhs) =>
>     Exception: not implemented - multiplication in modules over different rings (C++ exception)
>   exactreal::Element<exactreal::NumberField> exactreal::cppyy::mul(const exactreal::Element<exactreal::NumberField>& lhs, const exactreal::Element<exactreal::NumberField>& rhs) =>
>     Exception: not implemented - multiplication in modules over different rings (C++ exception)
>   exactreal::Element<exactreal::NumberField> exactreal::cppyy::mul(const exactreal::Element<exactreal::NumberField>& lhs, const exactreal::Element<exactreal::NumberField>& rhs) =>
>     Exception: not implemented - multiplication in modules over different rings (C++ exception)
>   exactreal::Element<exactreal::NumberField> exactreal::cppyy::mul(const exactreal::Element<exactreal::NumberField>& lhs, const exactreal::Element<exactreal::NumberField>& rhs) =>
>     Exception: not implemented - multiplication in modules over different rings (C++ exception)
> ```

Added a test for this as well.

**Comment by fingolfin** [comment]
@videlec @w-bruns I guess https://github.com/videlec/e-antic/issues/37 is the relevant issue regarding C++11 vs. newer C++ standards. I just commented there.

**Comment by videlec** [comment]
This is still failing for me
```
sage: x = polygen(QQ)
sage: K.<a> = NumberField(x^3 - 2 * x^2 - 3* x - 1, embedding=root)
sage: E = ExactReals(K)
sage: E(0) * E(1)
```

> > Serious bug
> > ```
> > sage: K.<sqrt2> = NumberField(x^2 - 2, embedding=AA(2).sqrt())
> > sage: E = ExactReals(K)
> > sage: E(0) * E(1)
> > ...
> > Exception: Could not instantiate mul<exactreal::Element<exactreal::NumberField>,exactreal::Element<exactreal::NumberField>>:
> >   none of the 5 overloaded methods succeeded. Full details:
> >   exactreal::Element<exactreal::NumberField> exactreal::cppyy::mul(const exactreal::Element<exactreal::NumberField>& lhs, const exactreal::Element<exactreal::NumberField>& rhs) =>
> >     Exception: not implemented - multiplication in modules over different rings (C++ exception)
> >   exactreal::Element<exactreal::NumberField> exactreal::cppyy::mul(const exactreal::Element<exactreal::NumberField>& lhs, const exactreal::Element<exactreal::NumberField>& rhs) =>
> >     Exception: not implemented - multiplication in modules over different rings (C++ exception)
> >   exactreal::Element<exactreal::NumberField> exactreal::cppyy::mul(const exactreal::Element<exactreal::NumberField>& lhs, const exactreal::Element<exactreal::NumberField>& rhs) =>
> >     Exception: not implemented - multiplication in modules over different rings (C++ exception)
> >   exactreal::Element<exactreal::NumberField> exactreal::cppyy::mul(const exactreal::Element<exactreal::NumberField>& lhs, const exactreal::Element<exactreal::NumberField>& rhs) =>
> >     Exception: not implemented - multiplication in modules over different rings (C++ exception)
> >   exactreal::Element<exactreal::NumberField> exactreal::cppyy::mul(const exactreal::Element<exactreal::NumberField>& lhs, const exactreal::Element<exactreal::NumberField>& rhs) =>
> >     Exception: not implemented - multiplication in modules over different rings (C++ exception)
> >   none of the 6 overloaded methods succeeded. Full details:
> >   exactreal::Element<exactreal::NumberField> exactreal::cppyy::mul(const exactreal::Element<exactreal::NumberField>& lhs, const exactreal::Element<exactreal::NumberField>& rhs) =>
> >     Exception: not implemented - multiplication in modules over different rings (C++ exception)
> >   exactreal::Element<exactreal::NumberField> exactreal::cppyy::mul(const exactreal::Element<exactreal::NumberField>& lhs, const exactreal::Element<exactreal::NumberField>& rhs) =>
> >     Exception: not implemented - multiplication in modules over different rings (C++ exception)
> >   exactreal::Element<exactreal::NumberField> exactreal::cppyy::mul(const exactreal::Element<exactreal::NumberField>& lhs, const exactreal::Element<exactreal::NumberField>& rhs) =>
> >     Exception: not implemented - multiplication in modules over different rings (C++ exception)
> >   exactreal::Element<exactreal::NumberField> exactreal::cppyy::mul(const exactreal::Element<exactreal::NumberField>& lhs, const exactreal::Element<exactreal::NumberField>& rhs) =>
> >     Exception: not implemented - multiplication in modules over different rings (C++ exception)
> >   exactreal::Element<exactreal::NumberField> exactreal::cppyy::mul(const exactreal::Element<exactreal::NumberField>& lhs, const exactreal::Element<exactreal::NumberField>& rhs) =>
> >     Exception: not implemented - multiplication in modules over different rings (C++ exception)
> >   exactreal::Element<exactreal::NumberField> exactreal::cppyy::mul(const exactreal::Element<exactreal::NumberField>& lhs, const exactreal::Element<exactreal::NumberField>& rhs) =>
> >     Exception: not implemented - multiplication in modules over different rings (C++ exception)
> > ```
> 
> Added a test for this as well.

This is still failing for me. As far as I can see there are only tests for `ExactReals()` not for `ExactReals(K)`.

**Comment by saraedum** [comment]
@videlec I only added the doctests, I have not pushed all the fixes yet.

**Comment by saraedum** [comment]
Making pickling pass was a bit complicated, as I need to hook into the JSON pickling that the C++ provides…but it seems to work now.

**Pull request merged**